src/gui/sdlmain.cpp

```diff
@@ -1411,27 +1424,84 @@ SDL_Window* GFX_GetWindow()
 
 DosBox::Rect GFX_GetCanvasSizeInPixels()
 {
-	const auto size = get_canvas_size_in_pixels(sdl.rendering_backend);
-	return to_rect(size);
+	return get_canvas_size_in_pixels(sdl.rendering_backend);
 }
 
 RenderingBackend GFX_GetRenderingBackend()
 {
 	return sdl.rendering_backend;
 }
 
-static SDL_Point restrict_to_viewport_resolution(const int w, const int h)
+static SDL_Rect get_desktop_size()
 {
-	if (sdl.viewport_resolution) {
-		return {std::min(iround(sdl.viewport_resolution->x * sdl.desktop.dpi_scale),
-		                 w),
-		        std::min(iround(sdl.viewport_resolution->y * sdl.desktop.dpi_scale),
-		                 h)};
-	} else {
-		return {w, h};
+	SDL_Rect desktop;
+	assert(sdl.display_number >= 0);
+
+	SDL_GetDisplayBounds(sdl.display_number, &desktop);
+
+	// Deduct the border decorations from the desktop size
+	int top    = 0;
+	int left   = 0;
+	int bottom = 0;
+	int right  = 0;
+
+	(void)SDL_GetWindowBordersSize(SDL_GetWindowFromID(sdl.display_number),
+	                               &top,
+	                               &left,
+	                               &bottom,
+	                               &right);
+	// If SDL_GetWindowBordersSize fails, it populates the values with 0.
+	desktop.w -= (left + right);
+	desktop.h -= (top + bottom);
+
+	assert(desktop.w >= FallbackWindowSize.x);
+	assert(desktop.h >= FallbackWindowSize.y);
+	return desktop;
+}
+
+static DosBox::Rect calc_restricted_viewport_size_in_pixels(const DosBox::Rect canvas_px)
+{
+	switch (sdl.viewport.mode) {
+	case ViewportMode::Fit:
+		if (sdl.viewport.fit.limit_rectangle) {
+			const auto viewport_px =
+			        sdl.viewport.fit.limit_rectangle->Scale(
+			                sdl.desktop.dpi_scale);
+
+			return canvas_px.Contains(viewport_px)
+			             ? viewport_px
+			             : viewport_px.ScaleToFit(canvas_px);
+
+		} else if (sdl.viewport.fit.desktop_percent) {
+			const auto desktop = get_desktop_size();
+			LOG_TRACE("*** 2: desktop: %d x %d", desktop.w, desktop.h);
+			const auto p = *sdl.viewport.fit.desktop_percent;
```

Comment:
If we programmatically expect the <optional> `desktop_percent` to be populated, suggest asserting before using it.

`assert(sdl.viewport.fit.desktop_percent)`

Reply:
That's what the `if (sdl.viewport.fit.desktop_percent)` does.